[202405] Add Cumulative CodeQL changes.

apop5 opened the pull request:
## Description

Cumulative CodeQL fixes for the release 202405 Branch.

13dff4ff03
f26f86dd85
e97c27388d
564f262817
769b019d1d
4828ebb596
4c372a273c
5f165d29f0
006dc1f1cd
c99d59f91f
f648481185
c70d96e19e
7c60e3309d
2deccbe69e
2c4db3932a
ec7a96c424
45219b51fa
b47eb1fb37
1cd17c14ed
fec2cd9e81
326dde3a9f
a420f6627f
4f4cc7d81a
d2e5518865
9496601f8f
83ec756f7c
192d343194
55ad1f28bd
e71970bff9
be03b5b183
b9d04a0817
f495191509
6acf82c3f5
2c6ee990f4
8398844583
f90d7a8ca3
5464d1661f
4c082e3c26
f8616ec497
070d25db52
442aab874c
fe7078e3e0
e0b29646ba
798cfa3b92
92dc249f0c
0c3eaacafc

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested
Ran Local CI.

## Integration Instructions
N/A

makubacki committed:
StandaloneMmPkg: CodeQL Fixes.

Makes changes to comply with alerts raised by CodeQL.

Most of the issues here fall into the following two categories:

1. Potential use of uninitialized pointer
2. Inconsistent integer width used in loop comparison

os-d commented on ShellPkg/Application/Shell/Shell.c:
```diff
@@ -2609,11 +2617,19 @@ RunCommandOrFile (
+        ASSERT (CommandWithPath != NULL);
+        ShellPrintHiiEx (-1, -1, NULL, STRING_TOKEN (STR_SHELL_NOT_FOUND), ShellInfoObject.HiiHandle, FirstParameter);
+        SetLastError (SHELL_NOT_FOUND);
+        SHELL_FREE_NON_NULL (CommandWithPath);
```
Shouldn't we not be doing this because CommandWithPath is NULL?

apop5 replied:
![image](https://github.com/user-attachments/assets/80ff73cd-7f76-42e1-b374-d6bedf5c19be)

I'll remove it.

os-d commented on ShellPkg/Application/Shell/ShellManParser.c:
```diff
@@ -624,8 +632,23 @@ ProcessManFile (
+        goto Done;
+      }
+
+      // MU_CHANGE End - CodeQL Change - unguardednullreturndereference
```
small nit: spurious END tag and the next START tag

os-d commented on ShellPkg/Application/Shell/ShellParametersProtocol.c:
```diff
@@ -1176,7 +1183,13 @@ UpdateStdInStdOutStdErr (
+          // MU_CHANGE Start - CodeQL Change - unguardednullreturndereference
+          if (TempHandle == NULL) {
+            ASSERT (TempHandle != NULL);
+            Status = EFI_OUT_OF_RESOURCES;
```
Looks like TempHandle doesn't get referenced again in this code path, but should we be returning here?

apop5 replied:
Returning here would require CommandLineCopy to be freed.

Looking at the code below, I think the error status will cause it to skip all the other code sections and reach the free statements.

If we bailed out here, we could miss the error messages that get displayed at the end of the function.

apop5 replied:
It does get used, it is assigned to StdErr, then its used internally.

os-d commented on ShellPkg/Application/Shell/ShellParametersProtocol.c:
```diff
@@ -1223,7 +1236,13 @@ UpdateStdInStdOutStdErr (
+            // MU_CHANGE Start - CodeQL Change - unguardednullreturndereference
+            if (TempHandle == NULL) {
+              ASSERT (TempHandle != NULL);
+              Status = EFI_OUT_OF_RESOURCES;
```
return?

apop5 replied:
same as above

os-d commented on ShellPkg/Application/Shell/ShellParametersProtocol.c:
```diff
@@ -1245,9 +1264,16 @@ UpdateStdInStdOutStdErr (
+        // MU_CHANGE Start - CodeQL Change - unguardednullreturndereference
+        if (TempHandle == NULL) {
+          ASSERT (TempHandle != NULL);
+          Status = EFI_OUT_OF_RESOURCES;
```
return?

apop5 replied:
same as above.

os-d commented on ShellPkg/Application/Shell/ShellParametersProtocol.c:
```diff
@@ -1262,9 +1288,16 @@ UpdateStdInStdOutStdErr (
+        // MU_CHANGE Start - CodeQL Change - unguardednullreturndereference
+        if (TempHandle == NULL) {
+          ASSERT (TempHandle != NULL);
+          Status = EFI_OUT_OF_RESOURCES;
```
return?

os-d commented on ShellPkg/Library/UefiHandleParsingLib/UefiHandleParsingLib.c:
```diff
@@ -309,25 +315,28 @@ LoadedImageProtocolDumpInformation (
+  // MU_CHANGE End - CodeQL Change - unguardednullreturndereference
   SHELL_FREE_NON_NULL (Temp);
   SHELL_FREE_NON_NULL (FilePath);
   SHELL_FREE_NON_NULL (CodeType);
```
We were just checking if these are null, we should also check before freeing them, right?

apop5 replied:
SHELL_FREE_NON_NULL checks if it is null, and does not free it in that case

os-d commented on ShellPkg/Library/UefiShellBcfgCommandLib/UefiShellBcfgCommandLib.c:
```diff
@@ -506,7 +519,12 @@ BcfgMod (
+      // MU_CHANGE End - CodeQL Change - unguardednullreturndereference
     }
 
     Status = EfiBootManagerLoadOptionToVariable (&LoadOption);
```
Should we return if DevicePath is NULL? What will happen if we try to boot a load option without a device path?

apop5 replied:
Please check updated Push

os-d commented on ShellPkg/Library/UefiShellDebug1CommandsLib/EditMenuBar.c:
```diff
@@ -115,7 +121,12 @@ MenuBarRefresh (
     FunctionKeyString = HiiGetString (gShellDebug1HiiHandle, Item->FunctionKeyToken, NULL);
+    // MU_CHANGE Start - CodeQL Change - unguardednullreturndereference
+    if (FunctionKeyString == NULL) {
+      return EFI_INVALID_PARAMETER;
```
Looks like this should free NameString first

os-d commented on ShellPkg/Library/UefiShellDebug1CommandsLib/EfiDecompress.c:
```diff
@@ -112,13 +122,19 @@ ShellCommandRunEfiDecompress (
+        if (ShellStatus == SHELL_SUCCESS) {
+          // MU_CHANGE End - CodeQL Change - unguardednullreturndereference
+          ASSERT (Temp64Bit <= (UINT32)(-1));
+          InSize = (UINTN)Temp64Bit;
```
Should this be with SafeIntLib instead?

apop5 replied:
Please check latest.

os-d commented on UefiCpuPkg/CpuMpPei/CpuMpPei.c:
```diff
@@ -517,7 +517,6 @@ InitializeExceptionStackSwitchHandlers (
   // MU_CHANGE [END] - CodeQL change
 
   SwitchStackData = (EXCEPTION_STACK_SWITCH_CONTEXT *)Buffer;
-
```
nit: can leave this

os-d commented on UefiCpuPkg/CpuMpPei/CpuMpPei.c:
```diff
@@ -553,14 +552,24 @@ InitializeMpExceptionStackSwitchHandlers (
-
+  // MU_CHANGE Start - CodeQL Change - unguardednullreturndereference
   if (EFI_ERROR (Status)) {
-    NumberOfProcessors = 1;
```
Looks like there already was an error case here, should we really be changing this?

os-d commented on UefiCpuPkg/CpuDxe/CpuMp.c:
```diff
@@ -761,25 +770,31 @@ InitializeMpSupport (
-  MpInitLibGetNumberOfProcessors (&NumberOfProcessors, &NumberOfEnabledProcessors);
-  mNumberOfProcessors = NumberOfProcessors;
-  DEBUG ((DEBUG_INFO, "Detect CPU count: %d\n", mNumberOfProcessors));
+  // MU_CHANGE Start - CodeQL Change - unguardednullreturndereference
```
Can you review this, looks like there was an upstream change here, not sure we didn't overwrite it

apop5 replied:
There was, and still is, no checking on the upstream MpInitLibGetNumberOfProcessors.

os-d commented on UefiCpuPkg/CpuMpPei/CpuBist.c:
```diff
@@ -177,125 +177,124 @@ CollectBistDataFromPpi (
+               (VOID **)&PlatformInformationRecord2
+               );
+    ASSERT_EFI_ERROR (Status);
+    PlatformInformationRecord2->NumberOfCpus = (UINT32)NumberOfProcessors;
```
Same here, can you review and make sure we didn't overwrite upstream change?

apop5 replied:
restored the upstream version

os-d commented on MdeModulePkg/Bus/Pci/PciBusDxe/PciEnumerator.c:
```diff
@@ -2021,14 +2072,12 @@ PciHotPlugRequestNotify (
+    return EFI_INVALID_PARAMETER;
+  } else if ((Operation == EfiPciHotplugRequestRemove) && (*NumberOfChildren != 0) && (ChildHandleBuffer == NULL)) {
+    return EFI_INVALID_PARAMETER;
+    // MU_CHANGE End - CodeQl Change
```
Looks unnecessary.

apop5 replied:
It is necessary, its just overly complex.

If trying to remove a hotplug device, but there are children, but the childbuffer is empty, its an invalid parameter.

os-d commented on MdeModulePkg/Bus/Usb/UsbBusDxe/UsbDesc.c:
```diff
@@ -1009,3 +1009,34 @@ UsbIoClearFeature (
+  EFI_USB_DEVICE_DESCRIPTOR  DevDesc;
+  UINT8                      Index;
+
+  // MU_CHANGE UNUSED VARIABLE
```
Should be rebased on to of Doug's recent change

os-d commented on MdeModulePkg/Core/Pei/Ppi/Ppi.c:
```diff
@@ -324,16 +324,20 @@ ConvertPpiPointersFv (
     for (GuidIndex = 0; GuidIndex < ARRAY_SIZE (GuidCheckList); ++GuidIndex) {
+      // MU_CHANGE Start - CodeQL Change
       //
       // Don't use CompareGuid function here for performance reasons.
```
This explicitly says don't use CompareGuid, are we sure we want to?

apop5 replied:
There was discussion around this when the codeql changes were first merged. 
Based on the discussion back then, this was an artifact, and we decided to just use the CompareGuid to satisfy codeql.

os-d reviewed (approved):
Reviewed as much as I could without my brain turning to mush

makubacki committed:
NetworkPkg: CodeQL Fixes.

Includes changes across the repo for the following CodeQL rules:
- cpp/comparison-with-wider-type
- cpp/overflow-buffer
- cpp/redundant-null-check-param
- cpp/uselesstest

Co-authored-by: Taylor Beebe <tabeebe@microsoft.com>
Co-authored-by: pohanch <125842322+pohanch@users.noreply.github.com>
Co-authored-by: kenlautner <85201046+kenlautner@users.noreply.github.com>

	ShellPkg: CodeQL Fixes.

Includes changes across the repo for the following CodeQL rules:
- cpp/comparison-with-wider-type
- cpp/overflow-buffer
- cpp/redundant-null-check-param
- cpp/uselesstest

Some false positives are filtered in CodeQlFilters.yml.

Co-authored-by: Taylor Beebe <31827475+TaylorBeebe@users.noreply.github.com>

makubacki committed:
MdePkg: CodeQL Fixes.

Includes changes across the repo for the following CodeQL rules:
- cpp/comparison-with-wider-type
- cpp/overflow-buffer
- cpp/redundant-null-check-param
- cpp/uselesstest

Co-authored-by: Taylor Beebe <tabeebe@microsoft.com>
Co-authored-by: kenlautner <85201046+kenlautner@users.noreply.github.com>

makubacki committed:
MdeModulePkg: CodeQL Fixes.
Includes changes across the repo for the following CodeQL rules:
- cpp/comparison-with-wider-type
- cpp/overflow-buffer
- cpp/redundant-null-check-param
- cpp/uselesstest

Co-authored-by: Taylor Beebe <tabeebe@microsoft.com>
Co-authored-by: pohanch <125842322+pohanch@users.noreply.github.com>
Co-authored-by: kenlautner <85201046+kenlautner@users.noreply.github.com>
Co-authored-by: Oliver Smith-Denny <osde@linux.microsoft.com>
Co-authored-by: Sean Brogan <sean.brogan@microsoft.com>
Co-authored-by: Aaron <105021049+apop5@users.noreply.github.com>

makubacki committed:
UefiCpuPkg: CodeQL Fixes.

Includes changes across the repo for the following CodeQL rules:
- cpp/comparison-with-wider-type
- cpp/overflow-buffer
- cpp/redundant-null-check-param
- cpp/uselesstest

Co-authored-by: kenlautner <85201046+kenlautner@users.noreply.github.com>

makubacki committed:
ShellPkg: CodeQL Fixes.

Includes changes across the repo for the following CodeQL rules:
- cpp/comparison-with-wider-type
- cpp/overflow-buffer
- cpp/redundant-null-check-param
- cpp/uselesstest

Some false positives are filtered in CodeQlFilters.yml.

Co-authored-by: Taylor Beebe <31827475+TaylorBeebe@users.noreply.github.com>

makubacki committed:
CodeQlFilters.yml: Add global CodeQL filter file to repo.

The CodeQL Analyze build plugin allows a global configuration file
to be specified for a repo that can define file patterns against
CodeQL rules to include and exclude results for the given file and
rule.

The global configuration file is useful in mu_basecore (vs specifying
the filters in individual package CI YAML files) because the repo
has high reuse. Integration repos can easily see all of the filter
activity in one file that is easy to inspect and include in their
build if needed.

apop5 commented:
Additional Files changed on top of 202311
ShellPkg\Library\UefiShellDebug1CommandsLib\EditMenuBar.c (missing freepool in MenuBarRefresh)
ShellPkg\Library\UefiShellDebug1CommandsLib\EfiDecompress.c (safeintlib in ShellCommandRunEfiDecompress)
ShellPkg\Library\UefiShellBcfgCommandLib\UefiShellBcfgCommandLib.c (handle out of resources case to exit)
